### DISCUSSION TITLE
Awesome WM support

### DISCUSSION THREAD

**Discussion by mnzaki** [infokiller/i3-workspace-groups#60] [Ideas]
I would like to write support for [AwesomeWM](awesomewm.org/) and I'm curious if there's interest in generalizing `i3-workspace-groups`, in general.
A cursory look at the code suggests that [I3Proxy](https://github.com/infokiller/i3-workspace-groups/blob/master/i3wsgroups/i3_proxy.py#L9) is (hopefully) the only interaction point with i3. I propose to abstract that into a `WMProxy` and `WMProxyAdapter`, rewrite `I3Proxy` as an adapter named `I3Adapter`, and write `AwesomeAdapter`. The scripts would be changed to `import WMProxy`.

Polybar support requires a similar effort, and I imagine folding that into `WMProxyAdapter`

Any input/guidance/suggestions/ideas?

> __Backstory__
> I've hacked together some workspace group management in my awesomewm config, inspired by `i3-workspace-groups`, to support functions of my ["activity" management script](https://github.com/mnzaki/mnzakish/blob/main/bin/activity).
>
> Now I'm jealous of the `i3-workspace-groups` code quality and feature set so I would like to join em'

**Comment by infokiller** [comment]
Hi @mnzaki and thanks for your interest!

I'm happy to generalize the project and merge PRs that add support for Awesome (or other WMs for that matter), with the following caveats:

1. Support for alternative WMs (not i3/sway) will be marked experimental, because I don't use them myself
2. Tests are required for most new code
3. There is some planned work that you need to make sure will be possible with awesome. You can review the open issues, and specifically look at https://github.com/infokiller/i3-workspace-groups/issues/57
4. The project name and many [scripts](https://github.com/infokiller/i3-workspace-groups/tree/master/scripts) have the `i3-` prefix in their name, which will make it confusing for non-i3 users. I guess it shouldn't be difficult to resolve, we could add `wm-` variants of these scripts or something, maybe at some point rename the project as well.

Let me know if this works for you.
